### PR TITLE
JWT Authentication: Add support for specifying groups in auth.jwt for teamsync

### PR DESCRIPTION
**Pull request by Jguer**
**What is this feature?**

pairing with @mgyongyosi 

- Refactor JWT settings
- Refactor JWT client
- Unify same functions across JWT and OAuth in utils
- Add support for specifying groups in auth.jwt for teamsync

### CODE REVIEW

File: pkg/setting/setting.go [hunk 1]
```diff
@@ -267,24 +267,7 @@ type Cfg struct {
...
-	JWTAuthKeyFile                 string
-	JWTAuthKeyID                   string
-	JWTAuthJWKSetFile              string
-	JWTAuthAutoSignUp              bool
-	JWTAuthRoleAttributePath       string
-	JWTAuthRoleAttributeStrict     bool
-	JWTAuthAllowAssignGrafanaAdmin bool
-	JWTAuthSkipOrgRoleSync         bool
-
+	JWTAuth AuthJWTSettings
```
Comment (linoman) [inline pkg/setting/setting.go:270]: so clean!

**Review by linoman — APPROVED** [review]
Great work! I really like the encapsulation of the JSON utils into the utils package!

File: pkg/services/authn/clients/jwt.go [hunk 2]
```diff
@@ -114,6 +112,14 @@ func (s *JWT) Authenticate(ctx context.Context, r *authn.Request) (*authn.Identi
 	id.OrgRoles = orgRoles
 	id.IsGrafanaAdmin = isGrafanaAdmin
 
+	if s.cfg.JWTAuth.GroupsAttributePath != "" {
+		groups, err := s.extractGroups(claims)
```
Comment (kalleep) [inline pkg/services/authn/clients/jwt.go:116]:
```suggestion
		groups, err := util.SearchJSONForStringSliceAttr(s.cfg.JWTAuth.GroupsAttributePath, claims)
```

Comment (kalleep) [reply pkg/services/authn/clients/jwt.go:116]: I dont think we need the extra function here. the only thing that one does is checking if GroupsAttributePath is empty and return and empty slice in that case. But we already do that check here

File: pkg/util/json.go [hunk 3]
```diff
@@ -1,4 +1,112 @@
...
+	// ErrFailedToUnmarshalJSON is an error for failure in unmarshalling JSON.
+	ErrFailedToUnmarshalJSON = errutil.NewBase(errutil.StatusBadRequest,
+		"json-failed-to-unmarshal", errutil.WithPublicMessage("Failed to unmarshal JSON"))
+
+	// ErrFailedToSearchJSON is an error for failure in searching JSON.
+	ErrFailedToSearchJSON = errutil.NewBase(errutil.StatusBadRequest,
+		"json-failed-to-search", errutil.WithPublicMessage("Failed to search JSON with provided path"))
+)
+
+// SearchJSONForGroups searches for groups in a JSON object and returns a string slice.
```
Comment (kalleep) [inline pkg/util/json.go:32]:
```suggestion
// SearchJSONForStringSliceAttr searches for slice attribute in a JSON object and returns a string slice.
```

**Review by kalleep — APPROVED** [review]
Great work! Something we may need to add is sync caching, same mechanism we have for auth proxy.

File: pkg/util/json.go (hunk 3 above)
Comment (Jguer) [reply pkg/util/json.go:32]: +1, need to also go cleanup the interface{} in the other func

File: pkg/util/json.go [hunk 4]
```diff
@@ -1,4 +1,112 @@
...
+	}
+
+	return "", nil
+}
+
+// searchJSONForAttr searches for a specific attribute in a JSON object.
+// The attributePath parameter is a string that specifies the path to the attribute.
+// The data parameter is the JSON object that we're searching.
+// The function returns the value of the attribute and an error if one occurred.
+func searchJSONForAttr(attributePath string, data interface{}) (interface{}, error) {
```
Comment (Jguer) [inline pkg/util/json.go:77]:
```suggestion
func searchJSONForAttr(attributePath string, data any) (any, error) {
```

File: pkg/services/authn/clients/jwt.go (hunk 2 above)
Comment (Jguer) [reply pkg/services/authn/clients/jwt.go:116]: if the attr is empty, it will return an err, prefer not to handle it

**Pull request merged**